mesonbuild/interpreterbase/decorators.py

```diff
@@ -554,6 +556,15 @@ def emit_feature_change(values: T.Dict[_T, T.Union[str, T.Tuple[str, str]]], fea
             for info in types:
                 types_tuple = info.types if isinstance(info.types, tuple) else (info.types,)
                 value = kwargs.get(info.name)
+                if isinstance(value, DefaultObject):
+                    # If we have a `default()` object passed to a required
+                    # argument give a more helpful error message than the
+                    # default. Otherwise, set the value to None so that we'll
+                    # go down the path to use the default value.
+                    if info.required:
+                        raise InvalidArguments(f'{name} got a default() value for the required keyword argument "{info.name}", default() is not valid for such an argument')
+                    value = None
```

Comment:
While I agree this is the more logical place to deal with DefaultObject, it also means it won't work for API that did not got ported to typed_kwarg yet. That will create very bad consistency issues and would be impossible to properly document. I think we need to handle this in the InterpreterBase layer, like we do for disabler AFAIK.

Reply:
Or we could just focus on getting @dcbaker's typed build targets PR in. Coming up on a year now.

Reply:
There are still some modules too, we would have to check everything got ported.